client/trino-client/src/main/java/io/trino/client/JsonResponse.java

```diff
@@ -108,37 +108,23 @@ public String toString()
                 .toString();
     }
 
-    public static <T> JsonResponse<T> execute(TrinoJsonCodec<T> codec, Call.Factory client, Request request, OptionalLong materializedJsonSizeLimit)
+    public static <T> JsonResponse<T> execute(TrinoJsonCodec<T> codec, Call.Factory client, Request request)
     {
         try (Response response = client.newCall(request).execute()) {
             ResponseBody responseBody = requireNonNull(response.body());
             if (isJson(responseBody.contentType())) {
-                String body = null;
                 T value = null;
                 IllegalArgumentException exception = null;
-                try {
-                    if (materializedJsonSizeLimit.isPresent() && (responseBody.contentLength() < 0 || responseBody.contentLength() > materializedJsonSizeLimit.getAsLong())) {
-                        // Parse from input stream, response is either of unknown size or too large to materialize. Raw response body
-                        // will not be available if parsing fails
-                        value = codec.fromJson(responseBody.byteStream());
-                    }
-                    else {
-                        // parse from materialized response body string
-                        body = responseBody.string();
-                        value = codec.fromJson(body);
-                    }
+                MaterializingReader reader = new MaterializingReader(responseBody.charStream(), 128 * 1024);
```

Comment:
This change probably regresses performance on queries not using the new client protocol, no? In the previous critical path here, Trino query results would always switch to chunked encoding on the server side which would prevent the `Content-Length` header from being sent and we would never materialize those results. Now we will always buffer the first 128KB of each response as well as going through the `Reader` interface instead of `InputStream` (which I believe is slower).

Reply:
We should decrease the buffer size I think. This is only used when exception occurs, right?

Reply:
The contents of the buffer will only be shown if the exception occurs, yeah- but the trick of the old logic was that the server wouldn't send a `Content-Length` on responses that it decided were too large and the client would always skip materializing those because of the `responseBody.contentLength() < 0` check. Reducing the buffer size might help, but I'm not sure how small you would have to go to avoid any noticeable performance difference since the previous implementation wouldn't allocate the buffer at all.